This notebook is focused on interacting with the Revelry API, which is an API for a social media platform. In the notebook, Python code is written that calls various endpoints of the Revelry API, retrieves data, and manipulates it in various ways. Please refer <a href='https://www.ravelry.com/api#index'>Revelry API</a> documentation for more details.

In [1]:
# import required libraries
import http.client
import json
import requests
import pandas as pd
import utils # helper class

In [2]:
# load API token username and password
# Note: Make sure to create your own username and password.
authUsername = '####'
authPassword = '####'

In [3]:
# define an instance of the class
api_instance = utils.HelperRaveleryAPIs(authUsername, authPassword)

In [4]:
# get color families
color_families = api_instance.get_color_families()
display(color_families)

color  id            name       permalink  spectrum_order
0   None   1          Yellow          Yellow               1
1   None   2   Yellow-orange   Yellow-orange               2
2   None   3          Orange          Orange               3
3   None   4      Red-orange      Red-orange               4
4   None   5             Red             Red               5
5   None   6      Red-purple      Red-purple               0
6   None   7          Purple          Purple               8
7   None   8     Blue-purple     Blue-purple               9
8   None   9            Blue            Blue              10
9   None  10      Blue-green      Blue-green              11
10  None  11           Green           Green              12
11  None  12    Yellow-green    Yellow-green              13
12  None  13           Black           Black              14
13  None  14           White           White              15
14  None  15            Gray            Gray              16
15  None  16           Brown           Brown              17
16  None  17            Pink            Pink              18
17  None  18  Natural/Undyed  Natural/Undyed              19
18  None  19    Multicolored    Multicolored              21
19  None  20       Rainbow\t         Rainbow              20

In [5]:
# get top patterns
top_hat_patterns = api_instance.get_patterns(query = 'hat', page = 15, page_size = 2) 
top_patterns = api_instance.get_patterns()
print("get_patterns columns: {}".format(top_patterns.columns))
print("top pattern names:" )
display(top_patterns['name'])

get_patterns columns: Index(['free', 'id', 'name', 'permalink', 'personal_attributes', 'first_photo',
       'designer', 'pattern_author', 'pattern_sources'],
      dtype='object')
top pattern names:


0                            Sophie Scarf
1                                  Valgus
2                       Celtic Rain Socks
3                                  Leicht
4     Twists & Turns: Westknits MKAL 2022
                     ...                 
95                   Anker's Summer Shirt
96                         Ingrid Sweater
97                   Dragons scales socks
98                Gnot Just Another Gnome
99                      Inclinations Cowl
Name: name, Length: 100, dtype: object

In [6]:
# get queued patterns
queued_patterns = api_instance.get_queue(rav_username = 'rieslingm', page_size = 1000)
print("queued_patterns columns: {}".format(queued_patterns.columns))
print("items in queue: {}".format(len(queued_patterns)))

queued_patterns columns: Index(['created_at', 'id', 'pattern_id', 'skeins', 'sort_order', 'user_id',
       'yarn_id', 'name', 'pattern_name', 'short_pattern_name', 'yarn_name',
       'pattern_author_id', 'pattern_author_name', 'best_photo', 'notes',
       'position_in_queue'],
      dtype='object')
items in queue: 958


In [7]:
# get favourites
favorite_patterns = api_instance.get_favorites(rav_username = 'rieslingm')
print("favorite_patterns columns: {}".format(favorite_patterns.columns))

403 Forbidden. This is not a read only API method.


JSONDecodeError: Extra data: line 1 column 5 (char 4)

Debugging:

Above code fails as it needs authenticated credentials and we have only read only credentials. 
Two errors received in above code:
1.  403: This is not read only API method and required authenticated credentials.
2.  JSONDecodeError: Due to assumption that output is specific format. Could be solved using try catch or a case statement.